File: src/main/java/it/unibo/crabinv/Controller/entities/enemy/EnemyController.java
package it.unibo.crabinv.Controller.entities.enemy;

import it.unibo.crabinv.Controller.core.audio.AudioController;
import it.unibo.crabinv.Controller.entities.entity.EntityAbstractController;
import it.unibo.crabinv.Controller.entities.entity.EntityNotCapableOfInputController;
import it.unibo.crabinv.Model.core.GameEngine;
import it.unibo.crabinv.Model.core.audio.SFXTracks;
import it.unibo.crabinv.Model.entities.enemies.Enemy;
import it.unibo.crabinv.Model.entities.entity.Delta;

import java.util.Random;

/**
 * It's the EnemyController, should control each enemy.
 */
public final class EnemyController extends EntityAbstractController<Enemy> implements EntityNotCapableOfInputController {
    private static final double ENEMY_SHOOTING_CHANCE = 0.007;
    private final AudioController audio;
    private final GameEngine engine;
    private final Random rand = new Random();

    /**
     * It's the enemyController, it needs the enemy, audio, and game engine to work.
     *
     * @param enemy the enemy to be created.
     * @param audio the audio that it's needed in the class.
     * @param engine the game engine.
     */
    public EnemyController(final Enemy enemy,
                           final AudioController audio,
                           final GameEngine engine) {
        super(enemy);
        this.audio = audio;
        this.engine = engine;
    }

    /**
     * Updates the status of the enemy.
     */
    @Override
    public void update(final Delta delta) {
        tick();
        move(delta);

        if (rand.nextDouble() < ENEMY_SHOOTING_CHANCE) {
            shoot();
        }
    }

    /**
     * Gives the speed of the enemy.
     *
     * @return the speed of the enemy
     */
    public double getSpeed() {
        return super.getEntity().getSpeed();
    }

    /**
     * Tells the enemy to go to a specific direction for 1 tick.
     *
     * @param delta either -1, 0 or 1, the former moves to the left, the latter moves to the right
     */
    private void move(final Delta delta) {
        super.getEntity().move(delta);
    }

    /**
     * Makes the enemy shoot if it can.
     */
    private void shoot() {
        if (super.getEntity().isAbleToShoot()) {
            super.getEntity().shoot();
            engine.spawnEnemyBullet(super.getEntity());
            audio.playSFX(SFXTracks.SHOT_ENEMY);
        }
    }

    /**
     * Updates the status for the tick.
     */
    private void tick() {
        super.getEntity().tick();
    }
}

File: src/main/java/it/unibo/crabinv/controller/entities/enemy/EnemyController.kt
package it.unibo.crabinv.controller.entities.enemy

import it.unibo.crabinv.controller.core.audio.AudioController
import it.unibo.crabinv.controller.entities.entity.AbstractEntityController
import it.unibo.crabinv.controller.entities.entity.EntityNotCapableOfInputController
import it.unibo.crabinv.model.core.audio.SFXTracks
import it.unibo.crabinv.model.core.engine.GameEngine
import it.unibo.crabinv.model.entities.enemies.Enemy
import it.unibo.crabinv.model.entities.entity.Delta
import java.util.Random

/**
 * It's the EnemyController, should control each enemy.
 * @param enemy the enemy to be created.
 * @param audio the audio that it's needed in the class.
 * @param engine the game engine.
 */
class EnemyController(
    enemy: Enemy?,
    private val audio: AudioController,
    private val engine: GameEngine,
) : AbstractEntityController<Enemy?>(enemy),
    EntityNotCapableOfInputController {
    private val rand = Random()

    override fun update(delta: Delta) {
        tick()
        move(delta)

        if (rand.nextDouble() < ENEMY_SHOOTING_CHANCE) {
            shoot()
        }
    }

    val speed: Double
        /**
         * Gives the speed of the enemy.
         * @return the speed of the enemy
         */
        get() = super.entity!!.speed

    /**
     * Tells the enemy to go to a specific direction for 1 tick.
     * @param delta either -1, 0 or 1, the former moves to the left, the latter moves to the right
     */
    private fun move(delta: Delta) {
        super.entity!!.move(delta)
    }

    /**
     * Makes the enemy shoot if it can.
     */
    private fun shoot() {
        if (super.entity!!.isAbleToShoot) {
            super.entity!!.shoot()
            engine.spawnEnemyBullet(super.entity)
            audio.playSFX(SFXTracks.SHOT_ENEMY)
        }
    }

    /**
     * Updates the status for the tick.
     */
    private fun tick() {
        super.entity!!.tick()
    }

    companion object {
        private const val ENEMY_SHOOTING_CHANCE = 0.007
    }
}
